text/0000-generic-int.md

```diff
@@ -0,0 +1,552 @@
+although the ability to set/get fields is still possible.
+
+For example, here's a modified version of our previous example:
+
+```rust
+#[repr(C, bitpacked)]
+struct MipsInstruction {
+    opcode: uint<6>,
+    rs: uint<5>,
+    rt: uint<5>,
+    immediate: u16,
+}
+```
+
+We could allow referencing `&self.immediate` as it is aligned to a byte
+boundary, although all of the other fields can only be copied (e.g. `self.rs`)
+or set explicitly (e.g. `self.rt = 4`). Because of this restriction, we would
+only allow `!Drop` types in bit-packed structs, as `Drop` requires a mutable
+reference. We would probably initially want to limit to `Copy` types, but I
+haven't thought much on why expanding to all `!Drop` types wouldn't be doable.
+
+## ASCII-specific methods for `uint<7>` and `[uint<7>]`
+
+All of the ASCII-specific methods for `u8` could be copied to `uint<7>`, where
+all values are valid ASCII. Additionally, longer-term, the potential for using
+`uint<7>` for ASCII-specific matching and splitting in `str` could be discussed.
+
+Additionally, casting from `[uint<7>]` to `str` could be made safe, as all ASCII
+strings are valid UTF-8. This could make code which interacts with ASCII much,
+much safer, without needing to do `from_utf8_unchecked` or risk the penalty of
+`from_utf8`.
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+This is a *big* change to the language, not to be taken lightly. Additionally,
+many libraries may start using `uint<N>` for non-power-of-two `N` because it
+reduces the number of invalid values, even if it's not the best solution. For
+example, a library taking a percentage from 0 to 100 might use a `uint<7>`,
+even though it really should just take a `u8` or `u32`.
+
+This requires const generics, which have not even been implemented in the
+compiler at the time of writing. This would be the first builtin type which uses
+generics at all, too.
+
+No other language provides this kind of generalisation, and it may be more effort
+on the compiler side than it's worth. For example, the existing 576 impls for
+`Shl` might just be better and more performant than four new generic ones.
+
+# Rationale and alternatives
+[alternatives]: #alternatives
+
+## Bound-based generalisation
+
+Generalising integers over their number of bits is certainly a very logical way
+to generalise integers *for computers*, but generalising based upon bounds is
+a very natural way for humans to do it, *and* more general. For example, instead
+of `uint<N>` and `int<N>` types, we could get away with just one type,
+`int<A..=B>`. This would be more powerful than the original: for example, a
+percentage could be represented exactly as `int<0..=100>`. Whether an integer is
+signed simply depends on whether its lower bound is negative.
+
+The primary reason for leaving this out is… well, it's a lot harder to
+implement, and could be added in the future as an extension. Longer-term, we
```

Comment:
Perhaps... tho you should elaborate on the implementation difficulty as you see it.

...but the ranges feel also much more useful generally for code that isn't interested in space optimizations and such things but rather want to enforce domain logic in a strict and more type-safe fashion. For example, you might want to represent a card in a deck as `type Rank = uint<1..=13>;`. Then you know by construction that once you have your `the_rank : Rank` then it is correct and you won't have to recheck things. Of course, the other, more elaborate type safe way is to use an `enum`, but it might also be less convenient to setup than a simple range.

I think this alternative should be seriously entertained as the way to go; then you can use type aliases / newtypes to map to the range based types, e.g. `type uint<const N: usize> = urange<{0..=pow(2, N)}>;`.

Reply:
You mean, `urange<{0..=pow(2, N) - 1}>`. :p 

But actually, you're right. I should seriously clarify that and write it down in the alternatives.